middleware/kubernetes/kubernetes.go

```diff
@@ -205,98 +223,125 @@ func (k *Kubernetes) Records(name string, exact bool) ([]msg.Service, error) {
-								}
-								records = append(records, s)
-							}
-						}
-					}
+		domain := svc.Name + "." + svc.Namespace + ".svc." + zone
+		if endpointName != "" {
+			domain = endpointName + "." + domain
+		}
+
+		servicesFound = true
+		if svc.Spec.ClusterIP != api.ClusterIPNone {
+			// This is a cluster service with a Cluster IP. Create a record for each matching port
+			for _, p := range svc.Spec.Ports {
+				if !portAndProtocolMatch(srvPort, p.Name, srvPortWildcard, p.Port, srvProtocol, string(p.Protocol), srvProtocolWildcard) {
+					continue
 				}
+				s := msg.Service{
+					Key:         msg.Path(strings.ToLower(domain), "coredns"),
+					Host:        svc.Spec.ClusterIP,
+					Port:        int(p.Port),
+					TargetStrip: 0}
+				records = append(records, s)
+
 			}
-		} else {
-			// Create records for each exposed port...
+			continue
+		}
+		// There was no ClusterIP, so this is a "headless" service.
+		// Find all endpoints for this service in the Endpoints cache
+		epList, _ := k.APIConn.epLister.List()
+		for _, ep := range epList.Items {
+			if !(ep.ObjectMeta.Name == svc.Name && ep.ObjectMeta.Namespace == svc.Namespace) {
+				continue
+			}
+			for _, eps := range ep.Subsets {
+				for _, port := range eps.Ports {
+					if !portAndProtocolMatch(srvPort, port.Name, srvPortWildcard, port.Port, srvProtocol, string(port.Protocol), srvProtocolWildcard) {
+						continue
+					}
+					for _, addr := range eps.Addresses {
+						// Add all endpoints if no endpoint name was given (as if wildcard)
+						// If a non wildcard endpoint name is given, only add that endpoint
+						if endpointName != "" && !symbolMatches(endpointName, getEndpointHostname(addr), epWildcard) {
+							continue
+						}
+						epHostname := getEndpointHostname(addr)
+						key := domain
+						if k.RequestType == "SRV" {
+							key = epHostname + "." + key
+						}
+						s := msg.Service{
+							Key:         msg.Path(strings.ToLower(key), "coredns"),
+							Host:        addr.IP,
+							Port:        int(port.Port),
+							TargetStrip: 0,
+						}
+						records = append(records, s)
+					}
 
-			for _, p := range item.Spec.Ports {
-				s := msg.Service{Key: msg.Path(strings.ToLower("_"+p.Name+"._"+string(p.Protocol)+"."+key), "coredns"), Host: clusterIP, Port: int(p.Port)}
-				records = append(records, s)
+				}
 			}
 		}
+
 	}
+	if len(records) == 0 && !servicesFound {
+		// Did not find any matching services
+		return nil, errNoItems
+	}
+	return records, nil
 
-	return records
 }
 
-// Get performs the call to the Kubernetes http API.
-func (k *Kubernetes) Get(namespace string, nsWildcard bool, servicename string, serviceWildcard bool, typeName string) ([]*api.Service, error) {
-	switch {
-	case typeName == "pod":
-		return nil, fmt.Errorf("pod not implemented")
-	default:
-		return k.getServices(namespace, nsWildcard, servicename, serviceWildcard)
+func getEndpointHostname(addr api.EndpointAddress) string {
```

Comment:
think we should just drop these `get` prefixes and follow go standard library; only use set or Set.